### ISSUE TITLE
termux-location: empty output

### ISSUE THREAD

**Issue by newkozlukov** [termux/termux-api#9]
Already mentioned in g+:
$ termux-location
yields empty output
$ termux-location -r last
{
"API_ERROR": "Failed to get location"
}﻿

Have same behaviour on two different devices both running CM12
// don't have environment for debugging yet

**Comment by Neo-Oli** [comment]
Are you perhaps trying this indoors? termux-location, by default, only yields any output if a precise GPS fix can be established. `termux-location -p network` will use less precise network triangulation but should be more reliable in providing an output.

**Comment by YEDrekt** [comment]
I still get no output whether I use `-p gps` or `-p network`. Only `-r last` works for me but since it always gives the same output, it doesn't have a use for me.

**Comment by reinout91** [comment]
I have the same problem, did you manage to solve it?

**Comment by newkozlukov** [comment]
> I have the same problem, did you manage to solve it?

Hello, @reintou91. Sorry, didn't really bother to. Just tried installing `termux-api` and running `termux-location` though, with positioning active for 5 minutes or so: observed same behaviour. No output.
My only guess is that there might be a timeout set somewhere that isn't handled properly and leads to silent termination of the script. I think so because in other navigation-related apps on my phone it takes a couple minutes of active searching to get even the first approximate location update.
However, from a quick glance at `termux-api.c` and reference for `LocationManager` I don't see where any possible timeout could be.
I also just tried checking this idea by running `termux-location -r updates` and I've been surprised to see that instead of continuously streaming location updates (even if it be an empty stream) the app successfully terminated after \approx 31 seconds of waiting (btw, `-r once` takes about a minute). This is because `REQUEST_UPDATES` line in `LocationApi.java` creates, for a reason I don't understand, a thread that `quit()`s the Looper after sleeping for 30 seconds. I don't think it should. Anyway, I can't check if waiting for an update for a longer time (i.e. running `-r updates` after removing that thread) would help as I don't have a ready android build environment.

Summing up:

- libc's `read()` isn't supposed to have any timeouts
- script terminates even if I recompile `termux-api.c` with `-W` option appended to `am`'s arguments
- `LocationManager` doesn't seem to support timeouts out of the box either
- I really don't get how loop quits in `REQUEST_ONCE`: it seems all lines lead either to json location or to an error. I get no output at all.
- Probably should open an issue for that 30 seconds thing in `-r updates`

**Comment by newkozlukov** [comment]
Apparently related to #98 #204

**Comment by richardxday** [comment]
I've experienced this and discovered what hopefully is a useful bit of information.

I leave termux running in the background all the time (no wake lock) and with a SSH server running.

This morning I ssh'd into termux (without unlocking the phone) and tried various forms of termux-location, all of them returned nothing after waiting a minute or two.

I then tried unlocking the phone and repeating the command (via SSH): no change

I then tried opening maps up (to ensure there was a location available) and repeating the command: no change

Then I brought the termux app to the foreground and typed the command into termux on the device. This then returned a location within a few seconds.

So I left the app in the foreground and tried the command via SSH (as before): this also worked and gave a location within a few seconds.

After that I locked the phone and tried the command via SSH again: no result.

Even with the phone unlocked but Termux not being in the foreground, the command does not work via SSH.

The only way I've got termux-location to work reliably is with the app in the foreground and in this situation, sending the command via SSH does work.

What was interesting was that I got a Termux: API Error, Error in ResultReturner notification at one point which points to an issue somewhere...

In summary I think that the app not being in the foreground is interfering with access to Android's location services, which kind of makes sense since the permissions for location talk about 'access whilst the app is running'.

**Comment by richardxday** [comment]
Just a useful further comment: I changed the location permission on Termux:API to 'allow all the time' and now termux-location works with the phone locked or the main app in the background.
